File: model/aether_model/phy/papr.py
"""Peak-to-average power ratio: measurement and reduction (roadmap P2-4).

OFDM sums 57 independently modulated carriers, so its envelope is very nearly complex
Gaussian and its peaks are large: this waveform measures 9–10 dB PAPR. That matters because
an SSB transmitter's PA is driven at a fixed *peak* — the operator sets drive until ALC just
stops acting — so every dB of PAPR is a dB of average power, and therefore of link margin,
thrown away. It is also what the field reports as "ugly ALC spikes"
(``docs/COMMUNITY-CONCERNS.md`` #8).

Two reduction techniques are implemented, both of which keep the ADR-0002 air interface
exactly as it is (a receiver needs no knowledge of either):

* :class:`ClipAndFilter` — clip the envelope, then re-apply the band-limiting filter to put
  back what the clipping splattered out of band, and iterate because filtering regrows the
  peaks. Costs nothing in throughput; pays in in-band distortion (EVM), which sets a ceiling
  on the SNR the highest modes can reach.
* :class:`ToneReservation` — sacrifice a few carriers to carry a peak-cancelling signal.
  Costs those carriers' payload; pays nothing in EVM, because the data carriers are never
  touched and the receiver simply ignores the reserved ones.

Both operate on the complex baseband envelope, which is the right domain: an SSB
transmitter's nonlinearity is memoryless in the envelope (see
:class:`~aether_model.channel.SaturatingPa`).
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from aether_model.phy.ofdm import carrier_map
from aether_model.phy.passband import band_limit_taps
from aether_model.waveform import WIDE_2300, WaveformParams

ComplexArray = NDArray[np.complex128]
FloatArray = NDArray[np.float64]


# ── measurement ───────────────────────────────────────────────────────


def papr_db(x: ComplexArray) -> float:
    """Peak-to-average power ratio of a burst, in dB."""
    p = np.abs(np.asarray(x)) ** 2
    mean = float(np.mean(p))
    if mean <= 0:
        return 0.0
    return float(10 * np.log10(np.max(p) / mean))


def ccdf(x: ComplexArray, levels_db: FloatArray | list[float]) -> FloatArray:
    """P(instantaneous power exceeds mean power by more than each level)."""
    p = np.abs(np.asarray(x)) ** 2
    rel = 10 * np.log10(np.maximum(p / np.mean(p), 1e-30))
    return np.array([float(np.mean(rel > lv)) for lv in np.asarray(levels_db)])


def out_of_band_db(
    x: ComplexArray, params: WaveformParams = WIDE_2300, guard_hz: float = 250.0
) -> float:
    """Power beyond the occupied band plus ``guard_hz``, relative to the power inside it.

    The guard keeps the filter's own transition skirt out of the measurement, so what is left
    is spectral regrowth — the thing that would land in a neighbour's passband.
    """
    x = np.asarray(x, dtype=np.complex128)
    spec = np.abs(np.fft.fft(x)) ** 2
    freq = np.fft.fftfreq(len(x), 1.0 / params.fs_baseband)
    edge = params.occupied_bandwidth_hz / 2 + guard_hz
    inside = float(np.sum(spec[np.abs(freq) <= params.occupied_bandwidth_hz / 2]))
    outside = float(np.sum(spec[np.abs(freq) > edge]))
    return float(10 * np.log10(max(outside, 1e-30) / max(inside, 1e-30)))


def evm_db(reference: ComplexArray, distorted: ComplexArray) -> float:
    """In-band error power after removing the best complex gain — a PA compresses, and a
    pure gain change is not distortion."""
    ref = np.asarray(reference, dtype=np.complex128)
    got = np.asarray(distorted, dtype=np.complex128)
    alpha = np.vdot(ref, got) / np.vdot(ref, ref)
    err = got - alpha * ref
    return float(10 * np.log10(np.mean(np.abs(err) ** 2) / np.mean(np.abs(alpha * ref) ** 2)))


def average_power_gain_db(x: ComplexArray, saturation_amplitude: float) -> float:
    """Average output power relative to the PA's saturation power — the number that actually
    sets link margin when drive is limited by peaks."""
    return float(10 * np.log10(np.mean(np.abs(x) ** 2) / saturation_amplitude**2))


CLIP_TARGET_DB = 5.0
"""Clip target for the constant-modulus modes (BPSK, QPSK, 8-PSK). Measured
(``bench/baselines/papr.csv``): EVM −22.8 dB, far more than they need — 8-PSK 1/2 decodes
20/20 at its threshold — and worth +1.0 … +1.7 dB of delivered power. These are the modes a
weak link actually runs on, so this is where the gain matters most."""
CLIP_TARGET_DENSE_DB = 7.0
"""Clip target for the amplitude-modulated constellations (16-QAM, 64-QAM). Their outer
points sit where the clipper works hardest: at the 5 dB target, 16-QAM 3/4 goes from 0 % to
45 % frame errors at its threshold. At 7 dB (EVM −31.9 dB) it is clean again, and the modes
still gain +0.25 … +0.50 dB."""


def clip_target_db(bits_per_symbol: int) -> float:
    """Clip target for a mode. Peak reduction is bought with in-band distortion, and how much
    a mode can absorb depends on whether its constellation carries information in amplitude:
    PSK rides through, QAM does not."""
    return CLIP_TARGET_DENSE_DB if bits_per_symbol >= 4 else CLIP_TARGET_DB


# ── reduction ─────────────────────────────────────────────────────────


def _clip(x: ComplexArray, threshold: float) -> ComplexArray:
    mag = np.abs(x)
    scale = np.ones_like(mag)
    hot = mag > threshold
    scale[hot] = threshold / mag[hot]
    return x * scale


class ClipAndFilter:
    """Iterative clip-and-filter (Armstrong, *Electron. Lett.* 2002).

    Each pass clips the envelope to ``target_papr_db`` above the mean and then runs the
    band-limiting filter, which removes the out-of-band splatter clipping created but also
    restores part of the peak — hence several passes. The filter is the same one the passband
    chain already applies, so nothing here can emit energy the transmitter would not.
    """

    def __init__(
        self,
        params: WaveformParams = WIDE_2300,
        target_papr_db: float = 6.0,
        iterations: int = 4,
    ) -> None:
        self.p = params
        self.target_papr_db = float(target_papr_db)
        self.iterations = int(iterations)
        self._taps = band_limit_taps(params)

    def _filter(self, x: ComplexArray) -> ComplexArray:
        # one pass of the *same* linear-phase FIR the passband chain uses, its constant group
        # delay taken out — exactly what a streaming transmitter does, so nothing here relies
        # on seeing the whole burst at once
        return np.asarray(np.convolve(x, self._taps, mode="same"), dtype=np.complex128)

    def process(self, x: ComplexArray) -> ComplexArray:
        y = np.asarray(x, dtype=np.complex128).copy()
        mean_power = float(np.mean(np.abs(y) ** 2))
        if mean_power <= 0:
            return y
        threshold = np.sqrt(mean_power * 10 ** (self.target_papr_db / 10))
        for _ in range(self.iterations):
            y = self._filter(_clip(y, threshold))
        # preserve average power so the comparison is like for like
        scale = np.sqrt(mean_power / max(float(np.mean(np.abs(y) ** 2)), 1e-30))
        return y * scale


class ToneReservation:
    """Peak cancellation on reserved carriers (Tellado, Stanford thesis, 1999).

    ``n_reserved`` data carriers are given up; the clipping noise is projected onto them and
    subtracted, so peaks come down without touching a single data symbol. The receiver needs
    no change — it already reads data only from the carriers the map assigns — but the
    reserved carriers no longer carry payload, which is the price.

    The reserved set is spread evenly across the band: peak cancellation needs the reserved
    tones to span the bandwidth, or the correction signal cannot be sharp in time.
    """

    def __init__(
        self,
        params: WaveformParams = WIDE_2300,
        n_reserved: int = 4,
        target_papr_db: float = 6.0,
        iterations: int = 8,
        step: float = 1.0,
    ) -> None:
        self.p = params
        self.cmap = carrier_map(params)
        data = np.asarray(self.cmap.data_carriers)
        if not 1 <= n_reserved <= len(data):
            raise ValueError("n_reserved must fit inside the data carriers")
        pick = np.linspace(0, len(data) - 1, n_reserved).round().astype(int)
        self.reserved_carriers = data[np.unique(pick)]
        self.n_reserved = len(self.reserved_carriers)
        self.target_papr_db = float(target_papr_db)
        self.iterations = int(iterations)
        self.step = float(step)
        self._bins = np.asarray(self.cmap.bins)[self.reserved_carriers] % params.fft_size

    @property
    def payload_cost(self) -> float:
        """Fraction of the data carriers given up."""
        return self.n_reserved / len(self.cmap.data_carriers)

    def process_symbol(self, x: ComplexArray) -> ComplexArray:
        """Reduce the peaks of one time-domain OFDM symbol body (``fft_size`` samples)."""
        y = np.asarray(x, dtype=np.complex128).copy()
        n = self.p.fft_size
        if len(y) != n:
            raise ValueError(f"expected one symbol body of {n} samples, got {len(y)}")
        threshold = np.sqrt(np.mean(np.abs(y) ** 2) * 10 ** (self.target_papr_db / 10))
        for _ in range(self.iterations):
            excess = y - _clip(y, threshold)
            if not np.any(excess):
                break
            spec = np.fft.fft(excess)
            kept = np.zeros_like(spec)
            kept[self._bins] = spec[self._bins]
            y = y - self.step * np.fft.ifft(kept)
        return y

File: core/aether-phy/src/papr.rs
//! Peak reduction for the transmitter (ADR-0004).
//!
//! OFDM adds 57 carriers with independent phases, so its envelope is close to Rayleigh and
//! its peaks run 10–12 dB above the mean. An amateur transmitter cannot pass that: driven to
//! a peak the rig will not clip, the *average* power — which is what carries the link — sits
//! more than 10 dB below what the finals can do. Trading a little distortion for a lower
//! peak-to-average ratio buys real transmit power, and on HF that is the difference between
//! a contact and a retry.
//!
//! # Clip and filter
//!
//! The method is Armstrong's (*Electronics Letters* 38(5), 2002): clip the envelope to a
//! target above the mean, then run the band-limiting filter, which removes the out-of-band
//! splatter clipping created — and, in doing so, restores part of the peak, which is why it
//! takes several passes. The filter is the *same* one the passband chain already applies, so
//! nothing here can emit energy the transmitter would not.
//!
//! # Why two targets
//!
//! Peak reduction is bought with in-band distortion, and how much a mode can absorb depends
//! on whether its constellation carries information in amplitude. This was measured the wrong
//! way round at first — tested at comfortable signal-to-noise ratios it looked as though only
//! 64-QAM cared. Re-measured at each mode's *own* threshold, 16-QAM 3/4 went from 0 % to 45 %
//! frame errors at the 5 dB target. So the split is by constellation, not by speed.

use crate::{
    Complex,
    fir::Fir,
    passband::band_limit_taps,
    waveform::{WIDE_2300, WaveformParams},
};

/// Clip target for the constant-modulus modes: BPSK, QPSK and 8-PSK.
///
/// Measured (`bench/baselines/papr.csv`): EVM −22.8 dB, far more than they need — 8-PSK 1/2
/// decodes 20 out of 20 at its threshold — and worth +1.0 … +1.7 dB of delivered power. These
/// are the modes a weak link actually runs on, so this is where the gain matters most.
pub const CLIP_TARGET_DB: f64 = 5.0;

/// Clip target for the amplitude-modulated constellations: 16-QAM and 64-QAM.
///
/// Their outer points sit where the clipper works hardest. At the 5 dB target, 16-QAM 3/4
/// goes from 0 % to 45 % frame errors at its threshold; at 7 dB (EVM −31.9 dB) it is clean
/// again, and the modes still gain +0.25 … +0.50 dB.
pub const CLIP_TARGET_DENSE_DB: f64 = 7.0;

/// How many clip-and-filter passes to run. Each one undoes part of the previous pass's peak
/// reduction, so the gain per pass falls away; four reaches the target on every mode.
pub const DEFAULT_ITERATIONS: usize = 4;

/// Clip target for a mode.
///
/// PSK rides through clipping; QAM does not, because its outer points are where the clipper
/// works hardest and its decisions depend on amplitude.
#[must_use]
pub fn clip_target_db(bits_per_symbol: usize) -> f64 {
    if bits_per_symbol >= 4 {
        CLIP_TARGET_DENSE_DB
    } else {
        CLIP_TARGET_DB
    }
}

/// Peak-to-average power ratio of a burst, in dB.
#[must_use]
pub fn papr_db(samples: &[Complex]) -> f64 {
    if samples.is_empty() {
        return 0.0;
    }
    let power = |&(re, im): &Complex| re.mul_add(re, im * im);
    let mean = samples.iter().map(power).sum::<f64>() / samples.len() as f64;
    if mean <= 0.0 {
        return 0.0;
    }
    let peak = samples.iter().map(power).fold(0.0, f64::max);
    10.0 * (peak / mean).log10()
}

/// In-band error power after removing the best complex gain, in dB.
///
/// A power amplifier compresses, and a pure gain change is not distortion; what is left after
/// the best-fit gain is taken out is what the receiver will see as noise.
///
/// # Panics
/// If the two bursts are different lengths, or the reference carries no power.
#[must_use]
pub fn evm_db(reference: &[Complex], distorted: &[Complex]) -> f64 {
    assert_eq!(reference.len(), distorted.len(), "bursts differ in length");
    // alpha = <ref, got> / <ref, ref>, the least-squares complex gain
    let mut cross = (0.0, 0.0);
    let mut energy = 0.0;
    for (&(rr, ri), &(gr, gi)) in reference.iter().zip(distorted) {
        cross.0 += rr.mul_add(gr, ri * gi);
        cross.1 += rr.mul_add(gi, -(ri * gr));
        energy += rr.mul_add(rr, ri * ri);
    }
    assert!(energy > 0.0, "the reference carries no power");
    let alpha = (cross.0 / energy, cross.1 / energy);

    let mut error = 0.0;
    let mut scaled = 0.0;
    for (&(rr, ri), &(gr, gi)) in reference.iter().zip(distorted) {
        let fitted = (
            alpha.0.mul_add(rr, -(alpha.1 * ri)),
            alpha.0.mul_add(ri, alpha.1 * rr),
        );
        error += (gr - fitted.0).mul_add(gr - fitted.0, (gi - fitted.1) * (gi - fitted.1));
        scaled += fitted.0.mul_add(fitted.0, fitted.1 * fitted.1);
    }
    10.0 * (error / scaled).log10()
}

/// Iterative clip-and-filter peak reduction.
#[derive(Debug, Clone)]
pub struct ClipAndFilter {
    /// Peak-to-average target, in dB.
    pub target_papr_db: f64,
    /// How many clip-then-filter passes to run.
    pub iterations: usize,
    taps: Vec<f64>,
}

impl ClipAndFilter {
    /// Build a peak reducer for a numerology.
    #[must_use]
    pub fn new(params: WaveformParams, target_papr_db: f64) -> Self {
        Self {
            target_papr_db,
            iterations: DEFAULT_ITERATIONS,
            taps: band_limit_taps(&params),
        }
    }

    /// The same reducer with a different pass count.
    #[must_use]
    pub fn with_iterations(mut self, iterations: usize) -> Self {
        self.iterations = iterations;
        self
    }

    /// One pass of the band-limiting filter with its constant group delay taken out.
    ///
    /// This is what a streaming transmitter does — the filter is already in the chain — so
    /// nothing here depends on seeing the whole burst at once.
    fn filter(&self, samples: &[Complex]) -> Vec<Complex> {
        let mut fir = Fir::<Complex>::new(self.taps.clone());
        let delay = fir.delay();
        let mut padded = samples.to_vec();
        padded.extend(std::iter::repeat_n((0.0, 0.0), delay));
        let filtered = fir.process(&padded);
        filtered[delay..].to_vec()
    }

    /// Reduce the peaks of a burst, keeping its average power.
    #[must_use]
    pub fn process(&self, samples: &[Complex]) -> Vec<Complex> {
        if samples.is_empty() {
            return Vec::new();
        }
        let power = |&(re, im): &Complex| re.mul_add(re, im * im);
        let mean_power = samples.iter().map(power).sum::<f64>() / samples.len() as f64;
        if mean_power <= 0.0 {
            return samples.to_vec();
        }
        let threshold = (mean_power * 10.0f64.powf(self.target_papr_db / 10.0)).sqrt();

        let mut out = samples.to_vec();
        for _ in 0..self.iterations {
            for sample in &mut out {
                let magnitude = sample.0.hypot(sample.1);
                if magnitude > threshold {
                    let scale = threshold / magnitude;
                    *sample = (sample.0 * scale, sample.1 * scale);
                }
            }
            out = self.filter(&out);
        }
        // restore the average power, so what changed is the peak and nothing else
        let after = out.iter().map(power).sum::<f64>() / out.len() as f64;
        let scale = (mean_power / after.max(1e-30)).sqrt();
        for sample in &mut out {
            *sample = (sample.0 * scale, sample.1 * scale);
        }
        out
    }
}

impl Default for ClipAndFilter {
    fn default() -> Self {
        Self::new(WIDE_2300, CLIP_TARGET_DB)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{modes::LONG, ofdm::OfdmModulator, waveform::WIDE_2300};

    /// A burst of OFDM symbols with deterministic pseudo-random carrier values, which is
    /// close enough to Gaussian for its envelope to behave like the real thing.
    fn ofdm_burst(symbols: usize) -> Vec<Complex> {
        let modulator = OfdmModulator::new(WIDE_2300);
        let n_data = WIDE_2300.n_data_carriers();
        let mut state: u64 = 0x2545_F491_4F6C_DD1D;
        let mut values: Vec<Vec<Complex>> = Vec::with_capacity(symbols);
        for _ in 0..symbols {
            let mut data: Vec<Complex> = Vec::with_capacity(n_data);
            for _ in 0..n_data {
                state ^= state >> 12;
                state ^= state << 25;
                state ^= state >> 27;
                let unit =
                    (state.wrapping_mul(0x2545_F491_4F6C_DD1D) >> 11) as f64 / (1u64 << 53) as f64;
                let angle = 2.0 * std::f64::consts::PI * unit;
                data.push((angle.cos(), angle.sin()));
            }
            values.push(modulator.data_symbol(&data));
        }
        modulator.modulate(&values)
    }

    #[test]
    fn the_targets_are_the_ones_adr_0004_settled_on() {
        assert!((clip_target_db(1) - CLIP_TARGET_DB).abs() < 1e-12, "BPSK");
        assert!((clip_target_db(2) - CLIP_TARGET_DB).abs() < 1e-12, "QPSK");
        assert!((clip_target_db(3) - CLIP_TARGET_DB).abs() < 1e-12, "8-PSK");
        assert!(
            (clip_target_db(4) - CLIP_TARGET_DENSE_DB).abs() < 1e-12,
            "16-QAM"
        );
        assert!(
            (clip_target_db(6) - CLIP_TARGET_DENSE_DB).abs() < 1e-12,
            "64-QAM"
        );
    }

    #[test]
    fn raw_ofdm_really_does_have_the_peaks_this_exists_to_remove() {
        let burst = ofdm_burst(LONG.data_symbols);
        let raw = papr_db(&burst);
        assert!(
            raw > 9.0,
            "the test signal is not representative: {raw:.1} dB"
        );
    }

    #[test]
    fn clipping_reaches_its_target() {
        let burst = ofdm_burst(LONG.data_symbols);
        for target in [CLIP_TARGET_DB, CLIP_TARGET_DENSE_DB] {
            let reduced = ClipAndFilter::new(WIDE_2300, target).process(&burst);
            let got = papr_db(&reduced);
            // the filter restores part of the peak on the last pass, so the target is
            // approached rather than met exactly; a dB of slack is what four passes leave
            assert!(
                got < target + 1.0,
                "target {target} dB, reached {got:.2} dB from {:.2} dB",
                papr_db(&burst)
            );
            assert!(got < papr_db(&burst) - 2.0, "it barely did anything");
        }
    }

    #[test]
    fn average_power_is_preserved() {
        // otherwise a comparison between clipped and unclipped is measuring the gain change
        let burst = ofdm_burst(8);
        let reduced = ClipAndFilter::default().process(&burst);
        let mean = |x: &[Complex]| {
            x.iter().map(|&(r, i)| r.mul_add(r, i * i)).sum::<f64>() / x.len() as f64
        };
        let (before, after) = (mean(&burst), mean(&reduced));
        assert!(
            (after / before - 1.0).abs() < 1e-9,
            "power moved by {:.3} dB",
            10.0 * (after / before).log10()
        );
    }

    #[test]
    fn the_distortion_it_costs_is_the_measured_amount() {
        // The number that decided ADR-0004: what a mode has to absorb in exchange for the
        // peak reduction. If this moves, the mode thresholds move with it.
        let burst = ofdm_burst(LONG.data_symbols);
        let five = ClipAndFilter::new(WIDE_2300, CLIP_TARGET_DB).process(&burst);
        let seven = ClipAndFilter::new(WIDE_2300, CLIP_TARGET_DENSE_DB).process(&burst);
        let (evm_five, evm_seven) = (evm_db(&burst, &five), evm_db(&burst, &seven));
        assert!(
            evm_five < -18.0,
            "5 dB target costs {evm_five:.1} dB of EVM, more than the modes were measured at"
        );
        assert!(
            evm_seven < evm_five,
            "the gentler target should distort less: {evm_seven:.1} vs {evm_five:.1}"
        );
    }

    #[test]
    fn an_empty_or_silent_burst_is_left_alone() {
        assert!(ClipAndFilter::default().process(&[]).is_empty());
        let silence = vec![(0.0, 0.0); 100];
        assert_eq!(ClipAndFilter::default().process(&silence), silence);
        assert!((papr_db(&silence) - 0.0).abs() < 1e-12);
    }

    #[test]
    fn evm_ignores_a_pure_gain_change() {
        // a compressing amplifier changes the gain; that is not distortion and must not be
        // counted as any
        let burst = ofdm_burst(4);
        let louder: Vec<Complex> = burst.iter().map(|&(r, i)| (0.5 * r, 0.5 * i)).collect();
        assert!(
            evm_db(&burst, &louder) < -200.0,
            "a gain change was counted as distortion: {}",
            evm_db(&burst, &louder)
        );
    }
}
